File: pythonModules/logtail.py
"""
Cloudflare Logger Client

This Python module provides a client class (`CloudflareLogger`) for interacting
with the `log-tail.hacolby.workers.dev` log service.

Features:
-   Convenience methods for logging (info, error, warn).
-   Batch log ingestion.
-   Log searching and AI analysis.
-   **Conditional Authentication:**
    -   Log ingestion endpoints (`/api/v1/logs/ingest*`) are public and
        do not send an API key.
    -   Search and Analysis endpoints (`/api/v1/logs/search`,
        `/api/v1/ai-analysis`) are secure and automatically send the
        `Authorization` header.
"""
import os
import requests
import json
from datetime import datetime
from typing import List, Dict, Any, Optional
from dotenv import load_dotenv

class CloudflareLogger:
    """
    A Python client for the Cloudflare Log Service.
    
    Manages sending logs, searching logs, and running AI analysis,
    applying authentication only when required.
    """
    
    def __init__(self, service_name: str, api_key: str):
        """
        Initializes the logger client.

        Args:
            service_name (str): The name of the service logging. This will
                                be attached to all log entries.
            api_key (str): The secret API key. This is *only* used for
                           secure endpoints (search, analysis) and is
                           *not* sent during log ingestion.
        """
        self.service_name = service_name
        self.api_key = api_key
        self.session = requests.Session()
        self.base_url = "https://log-tail.hacolby.workers.dev"

    def _endpoint_url(self, uri: str) -> str:
        """
        Creates the full, hardcoded URL for the log service API.
        Safely joins the base URL and the URI path, handling slashes.

        Args:
            uri (str): The API path (e.g., '/api/v1/test').
        
        Returns:
            str: The complete, combined URL.
        """
        clean_base = self.base_url.rstrip('/')
        clean_uri = uri.lstrip('/')
        return f"{clean_base}/{clean_uri}"

    def _request(
        self,
        method: str,
        path: str,
        json_payload: Optional[Dict[str, Any]] = None,
        params: Optional[Dict[str, Any]] = None,
        timeout: int = 10
    ) -> requests.Response:
        """
        A centralized helper for making API requests to the main log service.
        
        **Conditional Authentication:**
        This function automatically attaches the `Authorization` header
        *only* for non-ingestion endpoints (e.g., search, analysis).
        Ingestion endpoints are treated as public and do not send the key.
        """
        full_url = self._endpoint_url(path)
        
        headers = {
            'Content-Type': 'application/json'
        }

        # --- Conditional Authentication Logic ---
        # Only add the API key if the path is NOT for ingestion.
        is_ingestion = path.startswith('/api/v1/logs/ingest')
        if not is_ingestion:
            headers['Authorization'] = f'Bearer {self.api_key}'
        # ----------------------------------------

        try:
            response = self.session.request(
                method=method,
                url=full_url,
                headers=headers,
                json=json_payload,
                params=params,
                timeout=timeout
            )
            # Raise an HTTPError for bad responses (4xx or 5xx)
            response.raise_for_status()
            return response
        except requests.exceptions.HTTPError as http_err:
            print(f"[HTTP Error] {http_err.response.status_code} for {full_url}: {http_err.response.text}")
            raise
        except requests.exceptions.ConnectionError as conn_err:
            print(f"[Connection Error] Failed to connect to {full_url}: {conn_err}")
            raise
        except requests.exceptions.Timeout as timeout_err:
            print(f"[Timeout Error] Request to {full_url} timed out: {timeout_err}")
            raise
        except requests.exceptions.RequestException as req_err:
            print(f"[Request Error] An error occurred: {req_err}")
            raise

    def log(self, level: str, message: str, **metadata):
        """
        Sends a single log entry to the (public) ingestion endpoint.
        """
        payload = {
            'service_name': self.service_name,
            'level': level.upper(),
            'message': message,
            'timestamp': int(datetime.utcnow().timestamp() * 1000), # UTC ms
            'metadata': {
                **metadata,
                'timestamp_iso': datetime.utcnow().isoformat()
            }
        }
        
        try:
            self._request(
                'POST',
                '/api/v1/logs/ingest',
                json_payload=payload,
                timeout=5
            )
        except Exception as e:
            # Fallback to local print if service is unavailable
            print(f"[LOG ERROR] {e}: {message}")

    def batch_log(self, logs: List[Dict[str, Any]]):
        """
        Sends multiple logs at once to the (public) batch ingestion endpoint.
        
        Args:
            logs (List[Dict]): A list of log dictionaries. Each dict
                               should have 'level', 'message', 'metadata'.
        """
        processed_logs = [
            {
                'service_name': self.service_name,
                'level': log['level'].upper(),
                'message': log['message'],
                'timestamp': log.get('timestamp', int(datetime.utcnow().timestamp() * 1000)),
                'metadata': log.get('metadata', {})
            }
            for log in logs
        ]
        
        payload = {'logs': processed_logs}
        
        try:
            self._request(
                'POST',
                '/api/v1/logs/ingest/batch',
                json_payload=payload,
                timeout=10
            )
        except Exception as e:
            print(f"[BATCH LOG ERROR] {e}")

    # --- Convenience Methods ---

    def info(self, message: str, **metadata):
        """Logs an INFO level message."""
        self.log('INFO', message, **metadata)
    
    def error(self, message: str, **metadata):
        """Logs an ERROR level message."""
        self.log('ERROR', message, **metadata)
    
    def warning(self, message: str, **metadata):
        """Logs a WARN level message."""
        self.log('WARN', message, **metadata)
        
    def debug(self, message: str, **metadata):
        """Logs a DEBUG level message."""
        self.log('DEBUG', message, **metadata)

    # --- Secure Endpoints ---

    def search_logs(
        self,
        service: Optional[str] = None,
        level: Optional[str] = None,
        start_time: Optional[int] = None,
        end_time: Optional[int] = None,
        limit: int = 100,
        offset: Optional[int] = None,
        query: Optional[str] = None
    ) -> Dict[str, Any]:
        """
        Searches logs from the main logging service.
        This function calls a *secure* (auth-required) endpoint.

        Args:
            service (str, optional): Filter by service name.
            level (str, optional): Filter by log level.
            start_time (int, optional): UTC milliseconds start time.
            end_time (int, optional): UTC milliseconds end time.
            limit (int): Number of logs to return. Default 100.
            offset (int, optional): Offset for pagination.
            query (str, optional): Free-text search query.
        
        Returns:
            Dict[str, Any]: The JSON response from the search API
                            (e.g., {'logs': [...], 'total': ...}).
        """
        params = {
            'service': service,
            'level': level,
            'start_time': start_time,
            'end_time': end_time,
            'limit': limit,
            'offset': offset,
            'query': query
        }
        # Remove None values so they aren't in the query string
        clean_params = {k: v for k, v in params.items() if v is not None}
        
        try:
            response = self._request(
                'GET',
                '/api/v1/logs/search',
                params=clean_params
            )
            return response.json()
        except Exception as e:
            print(f"[SEARCH ERROR] {e}")
            return {'logs': [], 'error': str(e)}

    def analyze_logs(
        self,
        service: Optional[str] = None,
        start_time: Optional[int] = None,
        end_time: Optional[int] = None,
        search_keywords: Optional[str] = None
    ) -> str:
        """
        Performs AI analysis by calling the log service's built-in AI endpoint.
        This function calls a *secure* (auth-required) endpoint.

        Args:
            service (str, optional): Filter by service name.
            start_time (int, optional): UTC milliseconds start time.
            end_time (int, optional): UTC milliseconds end time.
            search_keywords (str, optional): Keywords for the AI to focus on.
        
        Returns:
            str: The string-based analysis from the service.
        """
        payload = {
            'service_name': service,
            'start_time': start_time,
            'end_time': end_time,
            'search_keywords': search_keywords
        }
        
        try:
            response = self._request(
                'POST',
                '/api/v1/ai-analysis',
                json_payload=payload
            )
            result = response.json()
            return result.get('analysis', 'No analysis content returned.')
        except Exception as e:
            print(f"[ANALYSIS ERROR] {e}")
            return f"Analysis failed: {e}"

# --- Usage Example ---
if __name__ == "__main__":

    # Load environment variables from .env file
    load_dotenv()

    # Fetch API key from .env
    api_key = os.getenv("LOGTAIL_API_KEY") or 'your-api-key'

    # Fetch Service Name from .env
    service_name = os.getenv("PYTHON_APP_NAME") or 'your-python-service-name'

    
    # API key is only used for search/analysis
    logger = CloudflareLogger(
        service_name=service_name or 'your-python-service-name',
        api_key=api_key or 'your-api-key' # No key needed if just logging
    )

    print("--- Sending Logs (No Auth Key Sent) ---")
    logger.info('Starting data processing', job_id='123', batch_size=1000)
    logger.error('Database connection failed', error_code='DB_TIMEOUT', retry_count=3)
    
    # Example of batch logging
    log_batch = [
        {'level': 'WARN', 'message': 'Disk space running low', 'metadata': {'percent_free': 10}},
        {'level': 'DEBUG', 'message': 'Processed item 456', 'metadata': {'item_id': 456}}
    ]
    logger.batch_log(log_batch)
    print("Logs sent.")

    print("\n--- Searching Logs (Auth Key Sent) ---")
    # Example: 1 hour ago
    start_ms = int((datetime.utcnow() - timedelta(hours=1)).timestamp() * 1000)
    
    try:
        search_results = logger.search_logs(
            service=service_name,
            level='ERROR',
            start_time=start_ms
        )
        print(f"Found {len(search_results.get('logs', []))} ERROR logs:")
        print(json.dumps(search_results, indent=2))
    except Exception:
        pass # Error already printed by _request

    print("\n--- Analyzing Logs (Auth Key Sent) ---")
    try:
        analysis = logger.analyze_logs(
            service=service_name,
            start_time=start_ms,
            search_keywords="timeout"
        )
        print("--- AI Analysis Result ---")
        print(analysis)
        print("--------------------------")
    except Exception:
        pass # Error already printed by _request

File: src/backendComponents/middleware/logtail.ts
/**
 * @file log_integration_worker.ts
 * @description Cloudflare Worker acting as a Log Integration Layer with a class-based client.
 *
 * Responsibilities:
 * 1) API Gateway (`fetch`): /search, /analyze, /ingest, /ingest-batch → forwards to log-tail service
 * 2) Log Consumer (`tail`): transforms tail events → batch ingest to log-tail service
 *
 * Design: LogIntegrationClient encapsulates env, base URL, service name resolution, conditional auth,
 *         and request helpers. Keeps fetch/tail handlers thin, stateless, and testable.
 */

// -----------------------------
// ENV & TYPES
// -----------------------------

interface Env {
  LOG_TAIL_CONSUMER: {
    tail(events: TailEvent[]): Promise<void>;
  };
  LOG_SERVICE_API_KEY: string;
}

type LogLevel = 'DEBUG' | 'INFO' | 'WARN' | 'ERROR' | 'CRITICAL';
type LogSourceType = 'http' | 'tail' | 'agent';

interface LogEntry {
  service_name: string;
  level: LogLevel;
  message: string;
  timestamp?: number;
  metadata?: Record<string, any>;
  source_type?: LogSourceType;
}

interface LogSearchParams {
  service?: string;
  level?: LogLevel;
  start_time?: number;
  end_time?: number;
  limit?: number;
  offset?: number;
  query?: string;
}

interface LogSearchResult {
  id: string;
  service_name: string;
  level: LogLevel;
  message: string;
  timestamp: number;
  metadata_json?: string;
  source_type: LogSourceType;
}

interface SearchResponse {
  logs: LogSearchResult[];
  total?: number;
  has_more?: boolean;
}

interface IngestResponse {
  success: boolean;
  id?: string;
  ingested?: number;
  failed?: number;
}

// -----------------------------
// UTILITIES
// -----------------------------

/** JSON response helper */
function jsonResponse(data: unknown, status = 200): Response {
  return new Response(JSON.stringify(data, null, 2), {
    status,
    headers: { 'Content-Type': 'application/json' },
  });
}

/** Extract subdomain (worker/service name) from a *.workers.dev URL */
function getServiceNameFromURL(requestUrl: string): string {
  try {
    const url = new URL(requestUrl);
    const parts = url.hostname.split('.');
    if (parts.length >= 3) {
      const [sub, secondLast, last] = parts.slice(-3);
      if (secondLast === 'workers' && last === 'dev') {
        return parts[0];
      }
    }
    return 'unknown-service';
  } catch {
    return 'unknown-service';
  }
}

// -----------------------------
// CLASS: LogIntegrationClient
// -----------------------------

class LogIntegrationClient {
  private readonly env: Env;
  private readonly baseUrl = 'https://log-tail.hacolby.workers.dev';
  public readonly serviceName: string;

  constructor(env: Env, requestUrl?: string, fallbackName = 'unknown-service') {
    this.env = env;
    this.serviceName = requestUrl ? getServiceNameFromURL(requestUrl) : fallbackName;
  }

  private endpoint(path: string): string {
    const cleanBase = this.baseUrl.replace(/\/$/, '');
    const cleanPath = path.replace(/^\//, '');
    return `${cleanBase}/${cleanPath}`;
  }

  private async request(path: string, init: RequestInit): Promise<Response> {
    const url = this.endpoint(path);
    const headers = new Headers(init.headers || {});
    headers.set('Content-Type', 'application/json');

    const isIngest = path.startsWith('/api/v1/logs/ingest');
    if (!isIngest) {
      headers.set('Authorization', `Bearer ${this.env.LOG_SERVICE_API_KEY}`);
    }

    const res = await fetch(url, { ...init, headers });
    if (!res.ok) {
      let msg = `Log Service Error (${res.status})`;
      try {
        const err = await res.json();
        msg = err.error || err.message || msg;
      } catch {}
      throw new Error(msg);
    }
    return res;
  }

  async ingestOne(entry: Omit<LogEntry, 'service_name'>): Promise<IngestResponse> {
    const logs = [{ ...entry, service_name: this.serviceName }];
    const res = await this.request('/api/v1/logs/ingest/batch', {
      method: 'POST',
      body: JSON.stringify({ logs }),
    });
    return res.json();
  }

  async ingestBatch(entries: Array<Partial<LogEntry> & { message: string; level: LogLevel }>): Promise<IngestResponse> {
    const logs = entries.map(e => ({
      service_name: e.service_name ?? this.serviceName,
      level: e.level,
      message: e.message,
      timestamp: e.timestamp,
      metadata: e.metadata,
      source_type: e.source_type,
    }));
    const res = await this.request('/api/v1/logs/ingest/batch', {
      method: 'POST',
      body: JSON.stringify({ logs }),
    });
    return res.json();
  }

  async search(params: LogSearchParams = {}): Promise<SearchResponse> {
    const qp = new URLSearchParams();
    const withDefaults = { service: this.serviceName, limit: 100, ...params };
    for (const [k, v] of Object.entries(withDefaults)) {
      if (v !== undefined && v !== null) qp.set(k, String(v));
    }
    const res = await this.request(`/api/v1/logs/search?${qp.toString()}`, { method: 'GET' });
    return res.json();
  }

  async analyze(params: LogSearchParams = {}): Promise<string> {
    const payload = {
      service_name: params.service ?? this.serviceName,
      start_time: params.start_time,
      end_time: params.end_time,
      search_keywords: params.query,
    };
    const res = await this.request('/api/v1/ai-analysis', {
      method: 'POST',
      body: JSON.stringify(payload),
    });
    const data = await res.json();
    return typeof data?.analysis === 'string' ? data.analysis : 'No analysis content returned.';
  }
}

// -----------------------------
// WORKER EXPORT
// -----------------------------

export default {
  async fetch(request: Request, env: Env, _ctx: ExecutionContext): Promise<Response> {
    const url = new URL(request.url);
    const path = url.pathname;
    const method = request.method;
    const client = new LogIntegrationClient(env, request.url);

    try {
      if (method === 'GET' && path === '/search') {
        const params: LogSearchParams = {
          level: (url.searchParams.get('level') as LogLevel) || undefined,
          query: url.searchParams.get('query') || undefined,
          limit: url.searchParams.get('limit') ? parseInt(url.searchParams.get('limit')!, 10) : undefined,
          start_time: url.searchParams.get('start_time') ? parseInt(url.searchParams.get('start_time')!, 10) : undefined,
          end_time: url.searchParams.get('end_time') ? parseInt(url.searchParams.get('end_time')!, 10) : undefined,
        };
        const results = await client.search(params);
        return jsonResponse(results);
      }

      if (method === 'POST' && path === '/analyze') {
        const body: LogSearchParams = await request.json();
        const analysis = await client.analyze(body);
        return jsonResponse({ analysis });
      }

      if (method === 'POST' && path === '/ingest') {
        const entry = (await request.json()) as Omit<LogEntry, 'service_name'>;
        const result = await client.ingestOne(entry);
        return jsonResponse(result);
      }

      if (method === 'POST' && path === '/ingest-batch') {
        const { logs } = (await request.json()) as { logs: Array<Partial<LogEntry> & { message: string; level: LogLevel }> };
        const result = await client.ingestBatch(logs);
        return jsonResponse(result);
      }

      return jsonResponse({ error: 'Not Found' }, 404);
    } catch (err: any) {
      console.error('Fetch error:', err?.stack || err?.message || err);
      return jsonResponse({ error: err?.message || 'Internal Server Error' }, 500);
    }
  },

  async tail(events: TailEvent[], env: Env, ctx: ExecutionContext): Promise<void> {
    const client = new LogIntegrationClient(env, undefined, 'log-integration');
    const entries: Array<Partial<LogEntry> & { message: string; level: LogLevel }> = [];

    for (const event of events) {
      try {
        let msg: any;
        if (typeof event.message === 'string') {
          try {
            msg = JSON.parse(event.message);
          } catch {
            msg = { message: event.message };
          }
        } else if (typeof event.message === 'object' && event.message) {
          msg = event.message;
        } else {
          msg = { message: 'Non-string/object tail message' };
        }

        entries.push({
          level: (msg.level || 'INFO') as LogLevel,
          message: msg.message || 'No message',
          timestamp: event.timestamp,
          metadata: { ...msg.metadata, tail_source: event.source, exception: event.exception },
          source_type: 'tail',
        });
      } catch (e: any) {
        console.error('Tail event parse error:', e?.message || e);
      }
    }

    if (entries.length) {
      ctx.waitUntil(
        client
          .ingestBatch(entries)
          .then(res => console.log(`Ingested ${res.ingested || 0} tail logs.`))
          .catch(err => console.error('Tail ingest error:', err?.stack || err?.message || err))
      );
    }
  },
};

// -----------------------------
// USAGE EXAMPLE (LOCAL TEST)
// -----------------------------

/**
 * Example local usage:
 *
 * const env = { LOG_SERVICE_API_KEY: 'your-key', LOG_TAIL_CONSUMER: { tail: async () => {} } };
 * const client = new LogIntegrationClient(env, 'https://my-worker.hacolby.workers.dev');
 *
 * // Send one log
 * await client.ingestOne({
 *   level: 'INFO',
 *   message: 'Startup complete',
 *   metadata: { version: '1.0.0', region: 'us-west1' },
 * });
 *
 * // Batch logs
 * await client.ingestBatch([
 *   { level: 'WARN', message: 'Disk space low', metadata: { freePercent: 8 } },
 *   { level: 'ERROR', message: 'Connection failed', metadata: { retry: true } },
 * ]);
 *
 * // Search logs
 * const results = await client.search({ query: 'ERROR', limit: 5 });
 * console.log(results.logs);
 *
 * // Analyze logs
 * const analysis = await client.analyze({ query: 'timeout' });
 * console.log('AI Analysis:', analysis);
 */
